File: Java/163-MissingRanges.java
public class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> result = new ArrayList<>();
        if (nums.length == 0){ //Empty array misses the range lower->upper.
            result.add(outputRange(lower, upper));
            return result;
        }
        int start = 0;
        if (nums[0] > lower){ //Handles lower boundary. Notice "inclusive".
            result.add(outputRange(lower, nums[0] - 1));
            start = nums[0];
        }else{
            start = lower;
        }
        for (int i = 0; i < nums.length; i++){
            if (nums[i] - start > 1){
                result.add(outputRange(start + 1, nums[i] - 1));//Misses range if distance > 1.
            }
            start = nums[i];
        }
        if (upper - start > 0){//handles upper bound
            result.add(outputRange(start + 1, upper));
        }
        return result;
    }
    
    private String outputRange(int lower, int upper){
        return (lower == upper ? lower + "" : lower + "->" + upper);
    }
    
}

//longer...
public class Solution {
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> res = new ArrayList<>();
        if (nums == null){
            return res;
        }
        if (nums.length == 0){
            if (lower == upper){
                res.add(Integer.toString(lower));
            }else{
                res.add(generateInterval(lower, upper));
            }
            return res;
        }
        
        if (nums[0] != lower){
            if (lower == nums[0] - 1){
                res.add(Integer.toString(lower));
            }else{
                res.add(generateInterval(lower, (nums[0] - 1)));
            }
        }
        int next = nums[0] + 1;
        
        for (int i = 1; i < nums.length; i++){
            if (nums[i] == nums[i-1]){
                continue;
            }
            if (nums[i] == next){
                next++;
                continue;
            }else{
                if (nums[i] - next == 1){
                    res.add(Integer.toString(next));
                }else{
                    res.add(generateInterval(next, nums[i] - 1));
                }
            }
            next = nums[i] + 1;
        }
        if (nums[nums.length - 1] != upper){
            System.out.println(upper + " " + next);
            if (nums.length == 1){
                if (nums[0] == upper - 1){
                    res.add(Integer.toString(upper));
                }else{
                    res.add(generateInterval(next, upper));
                }
            }else{
                if (upper == next){
                    res.add(Integer.toString(next));
                }else{
                    res.add(generateInterval(next, upper));
                }
            }
        }
        return res;
    }
    
    private String generateInterval(int lower, int upper){
        return lower + "->" + upper;
    }
}

File: cpp/163-MissingRanges.cpp
class Solution {
public:
    vector<string> findMissingRanges(vector<int>& nums, int lower, int upper) {
        vector<string> res;
        if (nums.size() == 0) {
            if (lower != upper)
                res.push_back(std::to_string(lower) + "->" + std::to_string(upper));
            else
                res.push_back(std::to_string(lower));
            return
                res;
        }
        int front = lower;
        int i = 0;
        for (i = 0; i < nums.size(); i++) {
            if (nums[i] > upper) {
                res.push_back(std::to_string(lower) + "->" + std::to_string(upper));
                break;
            }
            if (nums[i] == upper) {
                if (lower != upper) {
                    if (lower == upper - 1)
                        res.push_back(std::to_string(lower));
                    else
                        res.push_back(std::to_string(lower) + "->" + std::to_string(upper-1));
                }
                break;
            }
            if (nums[i] < lower)
                continue;
            if (nums[i] == lower) {
                lower++; continue;
            }
            if (nums[i] - 1 == lower) {
                res.push_back(std::to_string(lower));
            }
            else {
                res.push_back(std::to_string(lower) + "->" + std::to_string(nums[i]-1));
            }
            lower = nums[i] + 1;
        }
        if (i == nums.size()) {
            if (upper == nums[i-1] + 1)
                res.push_back(std::to_string(upper));
            else
                res.push_back(std::to_string(nums[i-1] + 1) + "->" + std::to_string(upper));
        }
        return res;
    }
};

//short version
class Solution {
public:
    string get_range(int start, int end)
    {
        return start==end? to_string(start) : to_string(start)+"->"+to_string(end);
    }
    vector<string> findMissingRanges(vector<int>& nums, int lower, int upper) {
        vector<string> result;
        int pre = lower-1;
        for(int i =0; i <= nums.size(); i++)
        {
           int cur = (i==nums.size()? upper+1:nums[i]);
           if(cur-pre>=2)
            result.push_back(get_range(pre+1,cur-1));
            pre = cur;
        }
        return result;
    }
};
